### PR TITLE
Purge Runtime Optimizations

### PR DESCRIPTION
**Pull request by ctsims**
Running a "Case Purge" with large databases was/is quite time consuming. This PR introduces some optimizations to improve the process

Current Performance Improvement on my phone is 8->10 seconds to 1.5-1.75 seconds

* Introduces the idea of metadata iteration for sqlstorage. 
* Add relationship to caseindex table, which can now provide all information about indexes that the casedb can. 
* Refactor purge to optimize for db read structure

cross-request: https://github.com/dimagi/commcare-core/pull/635

### CODE REVIEW

**Comment by amstone326** [comment]
@damagatchi retest this please

File: app/src/org/commcare/models/database/user/models/AndroidCaseIndexTable.java [hunk 1]
```diff
@@ -89,6 +92,73 @@ public void indexCase(Case c) {
         }
     }
 
+    public Vector<CaseIndex> getIndicesForCase(int recordId) {
```
[the remark this answers is not in the document]
Comment (ctsims) [reply app/src/org/commcare/models/database/user/models/AndroidCaseIndexTable.java]: It is currently unused, but I think I will need it fairly soon

**Comment by ctsims** [comment]
@amstone326 added the documentaiton requested.

**Comment by amstone326** [comment]
@wpride Ok I actually need to merge this bc I need to write a migration on another PR and it's going to be a mess if this isn't merged first. Can you still do a pass on it and if you see anything major, leave a note that can be addressed in a followup PR?

**Comment by wpride** [comment]
@ctsims @amstone326 looks all good to me

I'm assuming we'll want this in Formplayer as well ~ASAP?

File: app/src/org/commcare/engine/cases/CaseUtils.java [hunk 2]
```diff
@@ -105,4 +115,79 @@ public static void purgeCases() {
                 removedCaseCount, removedLedgers, taken));
 
     }
+
+    public static DAG<String, int[], String> getFullCaseGraph(SqlStorage<ACase> caseStorage,
```
Comment (wpride) [inline app/src/org/commcare/engine/cases/CaseUtils.java]: Would be nice to move these functions into `commcare-core`, maybe I'll take that on as part of the Formplayer PR

**Comment by ctsims** [comment]
Hey @wpride : Would be good to move the core DB operations into the formplayer compatible interfaces. I'm not sure whether purging takes ~any time at all on the server side, so possibly not worth worrying about moving over the full process (other than to keep a "one cannonical truth" purge algorithm).

**Pull request merged**